### PR TITLE
Lift toplevel values in classic mode

### PR DESCRIPTION
**Pull request by Keryan-dev**
Generalisation of #1151, reworked from #1171.

Couldn't help to do some unrelated cleaning.

### CODE REVIEW

File: middle_end/flambda2/from_lambda/closure_conversion.ml [hunk 1]
```diff
@@ -769,6 +768,37 @@ let close_named acc env ~let_bound_var (named : IR.named)
...
+  | Block_approximation (approx, _alloc_mode) -> (
+    let approx =
+      Simple.pattern_match field
+        ~const:(fun const ->
+          match Reg_width_const.descr const with
+          | Tagged_immediate i ->
+            let i = Targetint_31_63.to_int i in
+            if i >= Array.length approx
+            then
+              Misc.fatal_errorf
```
Comment (Ekdohibs) [inline middle_end/flambda2/from_lambda/closure_conversion.ml:790]: Isn't it possible for this to appear with GADT code.

Comment (lthls) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:790]: I haven't actually checked (I haven't got this branch checked out currently), but I would expect that with minimal changes you could get this to trigger.

File: middle_end/flambda2/from_lambda/closure_conversion.ml [hunk 2]
```diff
@@ -785,76 +811,184 @@ let close_let acc env id user_visible kind defining_expr
...
+                        :: fields)
+                    else None)
+                  f)
+            (Some []) fields
+          |> Option.map List.rev
+        in
+        match all_fields_static with
+        | Some static_fields ->
+          if (not (Flambda_features.classic_mode ()))
+             || not (Env.at_toplevel env)
```
[the remark this answers is not in the document]
Comment (Keryan-dev) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:864]: Right, we should allow toplevel constant block to be in the first branch and be syntactically lifted. I'll rework that.

**Review by lthls — COMMENTED** [review]
I think we can clean up/improve the comments, otherwise this looks fine. I'll do a last check that `CCenv.set_not_at_toplevel` is set where I expect it to, and after that we're good.

**Review by lthls — APPROVED** [review]

**Pull request merged**